lib/ansible/modules/cloud/amazon/ec2_eip.py

```diff
@@ -223,135 +223,172 @@
+                msg = "Couldn't associate Elastic IP address with network interface '{0}'".format(device_id)
+                module.fail_json_aws(e, msg=msg)
         if not res:
-            raise EIPException('association failed')
+            module.fail_json_aws(e, msg='Association failed.')
 
     return {'changed': True}
 
 
-def disassociate_ip_and_device(ec2, address, device_id, check_mode, is_instance=True):
-    if not address_is_associated_with_device(ec2, address, device_id, is_instance):
+def disassociate_ip_and_device(ec2, module, address, device_id, check_mode, is_instance=True):
+    if not address_is_associated_with_device(ec2, module, address, device_id, is_instance):
         return {'changed': False}
 
     # If we're in check mode, nothing else to do
     if not check_mode:
-        if address.domain == 'vpc':
-            res = ec2.disassociate_address(
-                association_id=address.association_id)
-        else:
-            res = ec2.disassociate_address(public_ip=address.public_ip)
-
-        if not res:
-            raise EIPException('disassociation failed')
+        try:
+            if address['Domain'] == 'vpc':
+                res = ec2.disassociate_address(
+                    AssociationId=address['AssociationId'],
+                    aws_retry=True
+                )
+            else:
+                res = ec2.disassociate_address(
+                    PublicIp=address['PublicIp'],
+                    aws_retry=True
+                )
+        except (botocore.exceptions.BotoCoreError, botocore.exceptions.ClientError) as e:
+            module.fail_json_aws(e, msg="Dissassociation of Elastic IP failed")
 
     return {'changed': True}
 
 
-def _find_address_by_ip(ec2, public_ip):
-    try:
-        return ec2.get_all_addresses([public_ip])[0]
-    except boto.exception.EC2ResponseError as e:
-        if "Address '{0}' not found.".format(public_ip) not in e.message:
-            raise
+def find_address(ec2, module, public_ip, device_id, is_instance=True):
+    """ Find an existing Elastic IP address """
+    filters = []
+    args = {}
 
+    if public_ip:
+        args["PublicIps"] = [public_ip]
+    elif device_id:
+        if is_instance:
+            filters.append({"Name": 'instance-id', "Values": [device_id]})
+        else:
+            filters.append({'Name': 'network-interface-id', "Values": [device_id]})
 
-def _find_address_by_device_id(ec2, device_id, is_instance=True):
-    if is_instance:
-        addresses = ec2.get_all_addresses(None, {'instance-id': device_id})
-    else:
-        addresses = ec2.get_all_addresses(None, {'network-interface-id': device_id})
-    if addresses:
-        return addresses[0]
+    if len(filters) > 0:
+        args["Filters"] = filters
+    elif len(filters) == 0 and public_ip is None:
+        return None
 
+    try:
+        addresses = ec2.describe_addresses(aws_retry=True, **args)
```

Comment:
The exception handling still works, but using aws_retry=True everywhere like this can have side effects as @willthames suggested. You can see this for yourself if you repeat the last task of the test suite a second time. Rather than the task failing immediately as it would do on devel, it retries for a long time and then finally fails (which is when the exception handling here gets to handle things).

As a side note, I would argue that using `state: absent` on something that doesn't exist should return `changed: False` instead of failure, but I know it's unrelated to your changes. The point is that the module should fail fast rather than slowly when possible.

So in this case, it would be better to add the retry decorator on find_address and handle the specific problematic error code inside this function, but re-raise other exceptions so the decorator can retry the operation. Something along these lines:
```
@AWSRetry.jittered_backoff()
def find_address(ec2, module, public_ip, device_id, is_instance=True):
    ...
    try:
        addresses = ec2.describe_addresses(**kwargs)
    except is_boto3_error_code('InvalidAddress.NotFound'):
        # Bypass retry logic by failing inside the retry decorator
        module.fail_json_aws(e, msg="Couldn't obtain list of existing Elastic IP addresses")
```
And then wherever find_address is called, catch the BotoCoreError and ClientError for retry failures.